### ISSUE TITLE
Nested lists of unequal length work as circuit parameters when given as lists but not as np.arrays 

### ISSUE THREAD

**Issue by cgogolin** [XanaduAI/pennylane#142]
This is a consequence of them being treated differently during (un)flatten() in util.py. I will look into this and suggest that we should either raise a more useful Exception or also deal with np.arrays of non-uniform row length.

**Comment by josh146** [comment]
Good catch. Let's update the exception now to be more useful in the meantime

**Comment by cgogolin** [comment]
It turns out that this is much more difficult than I thought: There is no easy way to test whether a nested `np.array()` is ragged or not. One can check the `dtype` to exclude that it is ragged, but if it is `object` then one has to recursively walk through all sub-arrays compare their length. This completely defeats the advantage of using the more efficient numpy way of (un)flattening such arrays, as the task of finding out whether fattening with these methods will work or not is almost es hard as (un)flattening them manually...

I thus thought it would be better to simply treat `np.array`s as `collection.Iterable`s (i.e., the same way `list`s  are treated (after all numpy arrays are also iterable). I thought this should be easy. Simply treat them exactly like `list`s and in the last step of each recursive call of `_unflatten()` convert back to `np.array`. However, even after spending more than two hours on this, I couldn't find a way of doing that, which at least made all the existing tests pass. I will look into this again next year...

**Comment by cgogolin** [comment]
I have made good progress getting to the root of this in the ` flatten_ragged_arrays` branch, but even with the improvements there, one can still not run the following simple program:

```
import pennylane as qml
from pennylane import numpy as np

num_wires = 1
dev = qml.device('default.qubit', wires=num_wires)

#a = [[1,2], np.array(3)] #this also doesn't work
a = [[1,2], 3]
a = np.array(a)#without this line everything works, obviously
print("a="+str(a))

@qml.qnode(dev)
def circuit(a):
    qml.Rot(a[0][0], a[0][1], a[1], wires=[0])
    return qml.expval.PauliZ(0)

print(circuit(a))
```

**Comment by cgogolin** [comment]
I think I got it fixed, except, probably, for special cases involving tuples. The code is quite an abomination, but once finishe it will come with extensive tests.

**Comment by cgogolin** [comment]
This will not be addressed in this release, but there is a partial solution and insights into what still needs to be fixed in https://github.com/XanaduAI/pennylane/tree/flatten_ragged_arrays_2

**Comment by cgogolin** [comment]
The problem with strings is now fixed in 399e62ff8ffdc1dc74d83d8a7134a60fca73eab7

**Comment by josh146** [comment]
Closing this issue for now, as it does not appear to occur in user interactions.
